exoscale/resource_exoscale_security_group_rules.go

```diff
@@ -129,6 +143,118 @@ func resourceSecurityGroupRules() *schema.Resource {
+	ICMPType              *uint8   `json:"icmp_type,omitempty"`
+	IDs                   []string `json:"ids,omitempty"`
+	Ports                 []string `json:"ports,omitempty"`
+	Protocol              string   `json:"protocol,omitempty"`
+	UserSecurityGroupList []string `json:"user_security_group_list,omitempty"`
+}
+
+func newStateSecurityGroupRuleFromInterface(rawStatePart interface{}) (*stateSecurityGroupRule, error) {
+	serializedRule, err := json.Marshal(rawStatePart)
+	if err != nil {
+		return nil, err
+	}
+
+	securityGroupRule := stateSecurityGroupRule{}
+	if err := json.Unmarshal(serializedRule, &securityGroupRule); err != nil {
+		log.Printf("[WARNING] %s", err)
+		return nil, err
+	}
+
+	return &securityGroupRule, nil
+}
+
+func (r stateSecurityGroupRule) toInterface() (map[string]interface{}, error) {
+	if len(r.UserSecurityGroupList) == 0 {
+		r.UserSecurityGroupList = nil
+	}
+
+	serializedRule, err := json.Marshal(r)
+	if err != nil {
+		return nil, err
+	}
+
+	var securityGroupRule map[string]interface{}
+	if err := json.Unmarshal(serializedRule, &securityGroupRule); err != nil {
+		return nil, err
+	}
+
+	return securityGroupRule, nil
+}
+
+func resourceSecurityGroupRulesStateUpgradeV0(_ context.Context, rawState map[string]interface{}, _ interface{}) (map[string]interface{}, error) {
+	log.Printf("[DEBUG] beginning migration")
+
+	// If we defined start_port to 0 with a previous version of the provider (< 0.31.x),
+	// the API backend will return start_port = 1.
+	// As a rule ID depends on its properties, in such a case, a rule ID doesn't match its definition anymore.
+	// Here we fix this by updating the rule ID from the current state.
+	var ruleIDRegex = regexp.MustCompile(`^([0-9a-z-]{36}_(?:tcp|udp)_.*)_0(-[0-9]+)?$`)
+	var rulePortsRegex = regexp.MustCompile(`^0-([0-9]+)$`)
+
+	for _, direction := range []string{"ingress", "egress"} {
+		if _, ok := rawState[direction]; !ok {
+			log.Printf("[DEBUG] flow direction not defined: '%s', skipping", direction)
+			continue
+		}
+
+		if rules, ok := rawState[direction].([]interface{}); ok {
+			patchRules := false
+			for idx, rule := range rules {
+				rule, err := newStateSecurityGroupRuleFromInterface(rule)
+				if err != nil {
+					return nil, err
+				}
+
+				// Fix rule IDs (start_port = 0 changed to 1)
+				for idx, ruleID := range rule.IDs {
+					rule.IDs[idx] = ruleIDRegex.ReplaceAllString(ruleID, "${1}_1${2}")
+					if ruleID != rule.IDs[idx] {
+						patchRules = true
+						log.Printf("[DEBUG] updated rule id from '%s' to '%s'\n", ruleID, rule.IDs[idx])
+					}
+				}
+
+				// Fix port range for the same reasons
+				for idx, ports := range rule.Ports {
+					rule.Ports[idx] = rulePortsRegex.ReplaceAllString(ports, "1-${1}")
+					if ports != rule.Ports[idx] {
+						patchRules = true
+						log.Printf("[DEBUG] updated rule ports from '%s' to '%s'\n", ports, rule.Ports[idx])
+					}
+				}
+
+				if patchRules {
```

Comment:
`patchRules ` should be set it back to `false` at the end of this statement, since we are in a loop here?

Reply:
Thank you :)